File: 00-Foundations/04-Design-Patterns/Behavioral/Observer/observer.ts
interface StockObserver {
  update(symbol: string, price: number): void;
}

class StockTicker {
  private observers: StockObserver[] = [];
  private price = 0;

  constructor(private readonly symbol: string) {}

  subscribe(observer: StockObserver): void {
    this.observers.push(observer);
  }

  unsubscribe(observer: StockObserver): void {
    this.observers = this.observers.filter((o) => o !== observer);
  }

  setPrice(price: number): void {
    this.price = price;
    this.observers.forEach((o) => o.update(this.symbol, this.price));
  }
}

class MobileAppDisplay implements StockObserver {
  update(symbol: string, price: number): void {
    console.log(`[Mobile App] ${symbol} is now ${price}`);
  }
}

class EmailAlert implements StockObserver {
  constructor(private readonly threshold: number) {}

  update(symbol: string, price: number): void {
    if (price >= this.threshold) {
      console.log(`[Email] Alert: ${symbol} crossed ${this.threshold}, now at ${price}`);
    }
  }
}

// Note: in real Node.js code, EventEmitter (from "events") already IS an
// Observer-pattern implementation — `emitter.on("price", cb)` /
// `emitter.emit("price", data)` is idiomatic and usually preferred over
// hand-rolling subscribe/notify, unless the interview specifically wants
// you to demonstrate the pattern from scratch.
function demo(): void {
  const ticker = new StockTicker("ACME");
  const mobile = new MobileAppDisplay();
  const alert = new EmailAlert(100);

  ticker.subscribe(mobile);
  ticker.subscribe(alert);

  ticker.setPrice(95);
  ticker.setPrice(105);
}

demo();

File: 00-Foundations/04-Design-Patterns/Behavioral/Observer/Observer.cs
namespace Foundations.Patterns.Behavioral.Observer;

public interface IStockObserver
{
    void Update(string symbol, decimal price);
}

public interface IStockSubject
{
    void Subscribe(IStockObserver observer);
    void Unsubscribe(IStockObserver observer);
}

// The subject: holds observers only through the shared interface, never
// concrete types.
public class StockTicker : IStockSubject
{
    private readonly List<IStockObserver> _observers = new();
    private readonly string _symbol;
    private decimal _price;

    public StockTicker(string symbol)
    {
        _symbol = symbol;
    }

    public void Subscribe(IStockObserver observer) => _observers.Add(observer);
    public void Unsubscribe(IStockObserver observer) => _observers.Remove(observer);

    public void SetPrice(decimal price)
    {
        _price = price;
        NotifyAll(); // push model: send the new state directly
    }

    private void NotifyAll()
    {
        foreach (var observer in _observers)
            observer.Update(_symbol, _price);
    }
}

public class MobileAppDisplay : IStockObserver
{
    public void Update(string symbol, decimal price) =>
        Console.WriteLine($"[Mobile App] {symbol} is now {price:C}");
}

public class EmailAlert : IStockObserver
{
    private readonly decimal _threshold;

    public EmailAlert(decimal threshold)
    {
        _threshold = threshold;
    }

    public void Update(string symbol, decimal price)
    {
        if (price >= _threshold)
            Console.WriteLine($"[Email] Alert: {symbol} crossed {_threshold:C}, now at {price:C}");
    }
}

public static class ObserverDemo
{
    public static void Run()
    {
        var ticker = new StockTicker("ACME");
        var mobile = new MobileAppDisplay();
        var alert = new EmailAlert(threshold: 100m);

        ticker.Subscribe(mobile);
        ticker.Subscribe(alert);

        ticker.SetPrice(95m);  // only mobile prints
        ticker.SetPrice(105m); // both print

        ticker.Unsubscribe(mobile);
        ticker.SetPrice(110m); // only alert prints
    }
}
